Analyses for “Experimental evolution makes microbes more cooperative
with the local host genotype”
================
Rebecca Batstone
2020-05-27

This repository archives the data and analyses for the paper:

Batstone RT, O’Brien AM, Harrison TL, Frederickson ME. Experimental
evolution makes microbes more cooperative with their local host
genotype.

[![DOI](https://zenodo.org/badge/267322857.svg)](https://zenodo.org/badge/latestdoi/267322857)

### Abstract

Advances in microbiome science require a better understanding of how
beneficial microbes adapt to hosts. We tested if hosts select for more
cooperative microbial strains with a year-long evolution experiment and
a cross-inoculation experiment designed to explore how nitrogen-fixing
bacteria (rhizobia) adapt to legumes. We paired *Ensifer meliloti* with
one of five *Medicago truncatula* genotypes that vary in how strongly
they ‘choose’ bacterial symbionts. Independent of host choice, *Ensifer
meliloti* rapidly adapted to its local host genotype and derived
microbes were more beneficial when they shared evolutionary history with
their host. This local adaptation was mostly limited to the symbiosis
plasmids, with mutations in putative signalling genes. Thus, cooperation
depends on the match between partner genotypes and increases as bacteria
adapt to their local host.

### Contents

This repository contains the raw data and analyses for:

##### Evolution experiment

This
[folder](https://github.com/rtbatstone/how-rhizobia-evolve/tree/master/Evolution%20experiment)
contains the raw data from the evolution experiment, the linear model
results presented in the main text of the paper, and the code that
generated Figure 1 in the main text and supplemental figure S6.

##### Single-inoculation experiment

This
[folder](https://github.com/rtbatstone/how-rhizobia-evolve/tree/master/Single-inoculation%20experiment)
contains the raw data from the single-inoculation experiment, the model
results presented in Tables S1 and S2, and the code that generated
Figure 2 in the main text, as well as supplemental figures S1 through
S4.

##### Variant discovery

This
[folder](https://github.com/rtbatstone/how-rhizobia-evolve/tree/master/Variant%20discovery)
contains the code and data to call and filter variants in the whole
genome sequences of our ancestral and derived *Ensifer meliloti*
strains. It also contains the code that generated supplemental figure
S5, and the code to calculate nucleotide diversity for ancestral versus
derived bacteria.

##### GWAS

This
[folder](https://github.com/rtbatstone/how-rhizobia-evolve/tree/master/GWAS)
contains the code to run the GWAS analyses and permutation tests
described in the paper, as well as the code to produce Figure 3 and the
information summarized in supplemental table S3.
